### DISCUSSION TITLE
Developing in Typescript

### DISCUSSION THREAD

**Discussion by DizzyHSlightlyVoided** [slint-ui/slint#6154] [Q&A] [answered]
Has anyone had any luck making Node-flavored Slint in Typescript? I've tried ...

```slint
/* ui/MainWindow.slint */
export component MainWindow inherits Dialog {
    /* ... */
}
```

```js
/* main.ts */
import * as slint from "slint-ui";

const ui = slint.loadFile('./ui/MainWindow.slint');
const mainWindow = new ui.MainWindow();
```
... but I immediately ran into an error: `Property 'MainWindow' does not exist on type 'Object'.`

**Comment by hunger** [comment]
@FloVanGH: Any ideas?

**Comment by ghost** [comment]
> Has anyone had any luck making Node-flavored Slint in Typescript? I've tried ...
> 
> ```slint
> /* ui/MainWindow.slint */
> export component MainWindow inherits Dialog {
>     /* ... */
> }
> ```
> 
> ```js
> /* main.ts */
> import * as slint from "slint-ui";
> 
> const ui = slint.loadFile('./ui/MainWindow.slint');
> const mainWindow = new ui.MainWindow();
> ```
> 
> ... but I immediately ran into an error: `Property 'MainWindow' does not exist on type 'Object'.`

Hi, on the first view it looks ok to me. Can you share a link with the whole code? Or can you sent paste the whole code of each file?

**Comment by DizzyHSlightlyVoided** [comment]
Here you go:

`package.json`
```json
{
  "name": "slint-helloworld",
  "version": "0.0.1",
  "main": "js/main.js",
  "type": "module",
  "scripts": {
    "start": "node ."
  },
  "dependencies": {
    "slint-ui": "^1.7.2"
  },
  "devDependencies": {
    "@types/node": "^22.5.5",
    "typescript": "^5.6.2"
  }
}
```

`tsconfig.json`
```json
{
  "compilerOptions": {
    "module": "ESNext",
    "target": "ESNext",
    "outDir": "js",
    "moduleResolution": "Node"
  },
  "include": [
    "./ts/**/*.ts"
  ]
}
```

`ts/main.ts
```ts
import * as slint from "slint-ui";

const ui = slint.loadFile('./ui/MainWindow.slint');
const mainWindow = new ui.MainWindow();

mainWindow.ok_clicked = mainWindow.hide;

await mainWindow.run();

process.exit();
```

`ui/MainWindow.slint`
```slint
import { StandardButton } from "std-widgets.slint";

export component MainWindow inherits Dialog {
    messageLabel := Text {
        text: "Hello, world!";
        wrap: word-wrap;
    }

    runButton := StandardButton {
        kind: ok;
    }
}
```

Notably, the Javascript itself works fine after running `npm install` and `npx tsc`, even though the Typescript has errors.

**Comment by ghost** [comment, marked as answer]
You have to add an `as any` to get it working:

```typescript
import * as slint from "slint-ui";

const ui = slint.loadFile('./ui/MainWindow.slint') as any;
const mainWindow = new ui.MainWindow();

mainWindow.ok_clicked = mainWindow.hide;

await mainWindow.run();

process.exit();
```

**Reply by DizzyHSlightlyVoided** [reply]
And of course, my immediate reaction was to put `as any` next to `new ui.MainWindow()` instead of next to `loadFile`. Thanks!
